src/features/dotnetTest.ts

```diff
@@ -35,41 +33,124 @@ export function registerDotNetTestDebugCommand(server: OmniSharpServer): vscode.
+            };
+        });
+}
+
+function getLaunchConfigurationForLaunch(server: OmniSharpServer, fileName: string, testMethod: string, testFrameworkName: string): Promise<any> {
+    const request: protocol.V2.GetTestStartInfoRequest = {
+        Filename: fileName,
+        MethodName: testMethod,
+        TestFrameworkName: testFrameworkName
+    };
+
+    return serverUtils.getTestStartInfo(server, request)
+        .then(response => {
+            let args = response.Argument.split(' ');
+
+            // Ensure that quoted args are unquoted.
+            args = args.map(arg => {
+                if (arg.startsWith('"') && arg.endsWith('"')) {
+                    return arg.substring(1, arg.length - 1);
+                }
+                else {
+                    return arg;
+                }
+            });
+
+            return {
+                name: ".NET Test Launch",
+                type: "coreclr",
+                request: "launch",
+                program: response.Executable,
+                args: args,
+                cwd: response.WorkingDirectory,
+                stopAtEntry: false
+            };
+        });
+}
+
+
+function getLaunchConfiguration(server: OmniSharpServer, debugType: string, fileName: string, testMethod: string, testFrameworkName: string): Promise<any> {
+    switch (debugType) {
+        case "attach":
+            return getLaunchConfigurationForAttach(server, fileName, testMethod, testFrameworkName);
+        case "launch":
+            return getLaunchConfigurationForLaunch(server, fileName, testMethod, testFrameworkName);
+
+        default:
+            throw new Error(`Unexpected PreferredDebugType: ${debugType}`);
+    }
+}
+
 // Run test through dotnet-test command with debugger attached
 export function debugDotnetTest(testMethod: string, fileName: string, testFrameworkName: string, server: OmniSharpServer) {
-    serverUtils.getTestStartInfo(server, { FileName: fileName, MethodName: testMethod, TestFrameworkName: testFrameworkName }).then(response => {
-        vscode.commands.executeCommand(
-            'vscode.startDebug', {
-                "name": ".NET test launch",
-                "type": "coreclr",
-                "request": "launch",
-                "program": response.Executable,
-                "args": response.Argument.split(' '),
-                "cwd": "${workspaceRoot}",
-                "stopAtEntry": false
+    const request: protocol.V2.DebugTestCheckRequest = {
+        Filename: fileName
+    };
+
+    let debugType: string;
+
+    return serverUtils.debugTestCheck(server, request)
+        .then(response => {
+            debugType = response.DebugType;
+            return getLaunchConfiguration(server, debugType, fileName, testMethod, testFrameworkName);
+        })
+        .then(config => {
+            return vscode.commands.executeCommand('vscode.startDebug', config);
+        })
+        .then(() => {
+            if (debugType === "attach") {
```

Comment:
Might want to add:
// TODO: Need to find out when the attach is really complete from the debugger. This is currently a race.